kliff/trainer/base_trainer.py

```diff
@@ -0,0 +1,636 @@
+
+    The core trainer class will provide the following functionality:
+    - Set up the work directory
+    - Set up the dataset
+    - Set up the test train split
+    Model, parameter transform and optimizer setup are left for the derived classes to
+    implement.
+
+    Args:
+        training_manifest: training manifest
+    """
+
+    def __init__(self, training_manifest: dict, model=None):
+        # workspace variables
+        self.workspace: dict = {
+            "name": "kliff_workspace",
+            "seed": 12345,
+            "resume": False,
+            "walltime": "2:00:00:00",
+        }
+
+        # dataset variables
+        self.dataset_manifest: dict = {
+            "type": "kliff",
+            "path": "./",
+            "save": False,
+            "shuffle": False,
+            "keys": {"energy": "energy", "forces": "forces"},
+            "colabfit_dataset": {
+                "dataset_name": None,
+                "database_name": None,
+                "database_url": None,
+            },
+        }
+        self.dataset = None
+
+        # model variables
+        self.model_manifest: dict = {
+            "type": "kim",
+            "name": None,
+            "path": None,
+        }
+        self.model: Callable = model
+
+        # transform variables
+        self.transform_manifest: dict = {
+            "property": [
+                {
+                    "name": None,
+                    "property_key": None,
+                }
+            ],
+            "parameter": [],
+            "configuration": {
+                "name": None,
+                "kwargs": None,
+            },
+        }
+        self.property_transforms = []
+        self.configuration_transform = None
+
+        # training variables
+        # this is too complicated to put it in singe dict, therefore the training
+        # block is divided into loss, optimizer, dataset_sample
+        self.training_manifest: dict = {}
+        self.loss_manifest: dict = {
+            "function": "mse",
+            "weights": {
+                "energy": 1.0,
+                "forces": None,
+                "stress": None,
+                "config": 1.0,
+            },
+            "normalize_per_atom": False,
+            "loss_traj": False,
+        }
+
+        self.optimizer_manifest: dict = {
+            "provider": "scipy",
+            "name": None,
+            "learning_rate": None,
+            "kwargs": None,
+            "epochs": 10000,
+            "stop_condition": None,
+            "num_workers": 1,
+            "batch_size": 1,
+        }
+        self.optimizer = None
+
+        # part of current?
+        self.dataset_sample_manifest: dict = {
+            "train_size": None,
+            "val_size": None,
+            "indices_files": {"train": None, "val": None},
+            "val_indices": None,
+            "train_indices": None,
+        }
+        self.train_dataset = None
+        self.val_dataset = None
```

Comment:
I did add it initially, but removed it as I did not see any use of it in MLIPs etc. I could not see that much of distinction between test and val sets. Can you tell where are you using `test` and `val` sets in you research? So that I can have some context